### PR TITLE
fix: [PL-38674]: Make delegate name unique within the scope in NG

### PR DESCRIPTION
**Pull request by jennyjames21**
## Describe your changes

Summary: 
Keep the delegate name unique within the same scope(project and organization)

Test
-Test with duplicate delegate names on different owners (different project ), Verified 2 delegates and 2 delegate groups getting created and both delegates showing under the corresponding scope in the UI
-Test with duplicate delegate names on account level and project level, Verified 2 delegate groups and delegates getting created and both delegates showing  under the corresponding scope in the UI
-Test duplicate names both under the same scope (both account level) and verified only one delegate group getting created and both delegates showing up under the same delegate group

## Checklist
- [x] I've documented the changes in the PR description.
- [x] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- 
  Specific builds
  - `trigger manager`
  - `trigger dms`
  - `trigger ng_manager`
  - `trigger cvng `
  - `trigger cmdlib`
  - `trigger template_svc`
  - `trigger events_fmwrk_monitor`
  - `trigger event_server`
  - `trigger change_data_capture`
  -  Trigger multiple builds together: For eg: `trigger dms, manager`
- Immutable delegate `trigger publish-delegate`
</details>

## [Latest PR Check Triggers](https://github.com/harness/harness-core/blob/develop/.github/pull_request_template.md)

<details>
  <summary>PR Check triggers</summary>
You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ut0, ut1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- Trigger CommonChecks: `trigger commonchecks`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger uts`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
- Validate_Reviews: `trigger review`
- Module Dependency Check: `trigger mdc`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/50870)

### CODE REVIEW

**Comment by jennyjames21** [comment]
trigger uts

**Comment by jennyjames21** [comment]
trigger uts

**Comment by jennyjames21** [comment]
Delegate with same name under project and account level
image
image
One delegate group
image
image

**Comment by jennyjames21** [comment]
trigger uts

**Comment by jennyjames21** [comment]
trigger uts

File: 400-rest/src/main/java/io/harness/delegate/service/impl/DelegateDownloadServiceImpl.java [hunk 1]
```diff
@@ -87,7 +87,8 @@ public DelegateDownloadResponse downloadDockerDelegate(String accountId, String
 
   private DelegateSetupDetails buildProperDelegateSetupDetails(String accountId, String orgIdentifier,
       String projectIdentifier, DelegateDownloadRequest delegateDownloadRequest, String delegateType) {
-    delegateService.checkUniquenessOfDelegateName(accountId, delegateDownloadRequest.getName(), true);
+    DelegateEntityOwner owner = DelegateEntityOwnerHelper.buildOwner(orgIdentifier, projectIdentifier);
```
Comment (raghuAtWings) [inline 400-rest/src/main/java/io/harness/delegate/service/impl/DelegateDownloadServiceImpl.java:90]: These code changes seem to be ensuring the uniqueness while downloading the yaml. Where is the logic to ensure delegate do get registered with the same name and different scope? 

Also i don't see any unit test on this change. Can we add one?

**Review by raghuAtWings — CHANGES_REQUESTED** [review]

File: 400-rest/src/main/java/io/harness/delegate/service/impl/DelegateDownloadServiceImpl.java (hunk 1 above)
Comment (jennyjames21) [reply 400-rest/src/main/java/io/harness/delegate/service/impl/DelegateDownloadServiceImpl.java:90]: DelegateServiceImpl line : 3209

Comment (jennyjames21) [reply 400-rest/src/main/java/io/harness/delegate/service/impl/DelegateDownloadServiceImpl.java:90]: There are already bunch of unit test on uniqueness check which is failing, i will fix and update that

**Comment by jennyjames21** [comment]
trigger uts

File: 400-rest/src/main/java/io/harness/delegate/service/impl/DelegateDownloadServiceImpl.java (hunk 1 above)
Comment (jennyjames21) [reply 400-rest/src/main/java/io/harness/delegate/service/impl/DelegateDownloadServiceImpl.java:90]: existing unit test

Comment (xingchi-jin) [reply 400-rest/src/main/java/io/harness/delegate/service/impl/DelegateDownloadServiceImpl.java:90]: yes, the line 3209 fixes the issue during registration.
and.. do we need to remove the delegateService.checkUniquenessOfDelegateName  in the download apis, and the k8s precheck api, these are UI functions for old installation wizard and CG, feel it's ok to keep them unchanged. And they will be removed by EoY .

Comment (jennyjames21) [reply 400-rest/src/main/java/io/harness/delegate/service/impl/DelegateDownloadServiceImpl.java:90]: Owner is passing as null for CG calls so no changes and yes they can be removed/unused once we fully adapt to the new wizard.

**Review by xingchi-jin — DISMISSED** [review]
LG, thanks !

**Comment by jennyjames21** [comment]
trigger uts

**Review by xingchi-jin — DISMISSED** [review]

**Comment by jennyjames21** [comment]
trigger uts

**Comment by jennyjames21** [comment]
trigger uts

**Review by raghuAtWings — APPROVED** [review]

**Pull request merged**